### ISSUE TITLE
A convenient way to apply PICARD on custom pytorch models?

### ISSUE THREAD

**Issue by atosystem** [ElementAI/picard#5]
I'm wondering if there is a convenient way to apply PICARD on my pytorch models while decoding?

Thanks!

**Comment by atosystem** [comment]
Thanks for your reply!
I am also doing the text-to-SQL task. However, I want to apply PICARD on my own seq2seq model.
Do I need to follow some interface? Or I can just simply change "model path" in  `config/server.json` to my model checkpoint?

**Comment by adityay121** [comment]
Hey @tscholak thank you for this awesome work, just a follow question, up on the comments made on this thread, I am working on this problem, where an end user will input a query in English, as per my understanding,  the inputs of your model are supposed to be in a particular format, which is as below. My question is mode of an engineering 
```    [question] | [db_id] | [table] : [column] ( [content] , [content] ) , [column] ( ... ) , [...] | [table] : ... | ... ```
You can't really expect an end user to input the query in this said format, the best they would be able to do is to do is input the [question].

Are there are anyways that you would suggest the to extract the rest of the prompts (`[db_id]` ,'`[column`') from the `[question]`?

**Comment by tscholak** [comment]
Hi @adityay121,
Great question! Have a look at [this example](https://huggingface.co/tscholak/3vnuv1vf?text=How+many+singers+do+we+have%3F+%7C+concert_singer+%7C+stadium+%3A+stadium_id%2C+location%2C+name%2C+capacity%2C+highest%2C+lowest%2C+average+%7C+singer+%3A+singer_id%2C+name%2C+country%2C+song_name%2C+song_release_year%2C+age%2C+is_male+%7C+concert+%3A+concert_id%2C+concert_name%2C+theme%2C+stadium_id%2C+year+%7C+singer_in_concert+%3A+concert_id%2C+singer_id):
<img width="1008" alt="Screen Shot 2021-11-24 at 3 27 05 PM" src="https://user-images.githubusercontent.com/1568873/143308731-d37ed048-d7d7-47de-a942-1544c9e16101.png">
Questions are always grounded in a database schema. In this example, the schema is called `concert_singer`, and the SQL commands to create it are:
```sql
CREATE TABLE "stadium" (
  "Stadium_ID" int,
  "Location" text,
  "Name" text,
  "Capacity" int,
  "Highest" int,
  "Lowest" int,
  "Average" int,
  PRIMARY KEY ("Stadium_ID")
);

CREATE TABLE "singer" (
  "Singer_ID" int,
  "Name" text,
  "Country" text,
  "Song_Name" text,
  "Song_release_year" text,
  "Age" int,
  "Is_male" bool,
  PRIMARY KEY ("Singer_ID")
);

CREATE TABLE "concert" (
  "concert_ID" int,
  "concert_Name" text,
  "Theme" text,
  "Stadium_ID" text,
  "Year" text,
  PRIMARY KEY ("concert_ID"),
  FOREIGN KEY ("Stadium_ID") REFERENCES "stadium"("Stadium_ID")
);

CREATE TABLE "singer_in_concert" (
  "concert_ID" int,
  "Singer_ID" text,
  PRIMARY KEY ("concert_ID","Singer_ID"),
  FOREIGN KEY ("concert_ID") REFERENCES "concert"("concert_ID"),
  FOREIGN KEY ("Singer_ID") REFERENCES "singer"("Singer_ID")
);
```
You can find a "linearization" of this schema after the first separator, `|`, in the input to the model. The first field is the database id, `concert_singer`, and after that we see a list of all table names and all column names.

To answer your question directly: The rest of the prompt is not extracted from the question ("How many singers do we have?"). Instead, that part of the input is assembled entirely from the SQL schema. It is the model's task to interpret the question in the context of the schema (and vice versa). You do not have to do that. But you have to know the schema of your target database!

This repository provides code that prepares the input to the model in the correct form from user-provided questions and `.sqlite` files. It is all set up and ready to be used, just start a server with the `make serve` command. See https://github.com/ElementAI/picard#serving for more details.

**Comment by adityay121** [comment]
Hi @tscholak If I wanted to use my own Seq-2-Seq t-5 base/large models which have been trained on a new dataset, consisting of SQL subqueries (not present on the original spider dataset) would changing the HF checkpoints on configs suffice or do I need to make any more changes?

**Comment by tscholak** [comment]
Hi, can you give an example of what you mean by a "subquery"?

**Comment by adityay121** [comment]
yeah sure, something like this ```SELECT LEFT(sub.date, 2) AS cleaned_month,
       sub.day_of_week,
       AVG(sub.incidents) AS average_incidents
  FROM (
        SELECT day_of_week,
               date,
               COUNT(incidnt_num) AS incidents
          FROM tutorial.sf_crime_incidents_2014_01
         GROUP BY 1,2
       ) sub
 GROUP BY 1,2
 ORDER BY 1,2```
you might as well refer to https://mode.com/sql-tutorial/sql-sub-queries/

**Comment by tscholak** [comment]
Hi @adityay121, There are nested SQL queries in the Spider dataset, I would not be worried about that. You can try to use your model with this code and see if it works.

**Comment by tsdocode** [comment]
Hi @tscholak , Is it possible to replace **T5** model with **GPT-Neo**?

**Comment by tscholak** [comment]
Hi @tsdocode! In theory, yes. GPT-J may also work. I'm interested in trying this myself, perhaps we can work on that together?

**Comment by tsdocode** [comment]
@tscholak yah, I tried fine-tuning the GPT-NEO 125M model on wikisql to your input format and got a better result, so I'm going to use PICARD on that model before moving on to other GPT models.

**Comment by tscholak** [comment]
Hi all, I'm closing this issue due to inactivity. Feel free to reopen if you'd discuss this again.
